### PR TITLE
SoapySDR support

### PR DESCRIPTION
**Pull request by bkoehm**
- Discussion in issue #150.
- Adds to original work by Oliver Jowett, January 2019.

### CODE REVIEW

File: sdr_soapy.c [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            printf("%s=%s", result.keys[i], result.vals[i]);
+            if (i+1 < result.size) printf(", ");
+        }
+        printf("\n");
+    }
+    SoapySDRKwargs_clear(&result);
+
+    char* hw_key = SoapySDRDevice_getHardwareKey(SOAPY.dev);
+    if (hw_key) {
+        printf("soapy: hardware key is %s\n", hw_key);
```
Comment (mutability) [inline sdr_soapy.c:128]: What's this actually used for? I couldn't find references to hw_key elsewhere.

File: sdr_soapy.c [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        if (SOAPY.dev_sdrplay) {
+            gain = Modes.gain;
+        } else {
+            gain = Modes.gain / 10.0;
+        }
+        if (SoapySDRDevice_setGain(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, gain) != 0) {
+            fprintf(stderr, "soapy: setGain failed: %s\n", SoapySDRDevice_lastError());
+            goto error;
+        }
+    }
```
Comment (mutability) [inline sdr_soapy.c:193]: This should probably unconditionally use Modes.gain directly with no division.

In older dump1090 versions, Modes.gain was an integer in tenths-of-a-dB. In more recent versions, it's just directly a floating-point gain in dB. I assume that SoapySDR wants gain in floating-point-dB (haven't looked at that in a while, but that's what I recall anyway)

File: sdr_soapy.c [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+    if (SoapySDRDevice_setBandwidth(SOAPY.dev, SOAPY_SDR_RX, SOAPY.channel, (SOAPY.dev_sdrplay ? 5.0e6 : 3.0e6)) != 0) {
```
Comment (mutability) [inline sdr_soapy.c:195]: I'd prefer to make bandwidth configurable as a command-line option, and then set _defaults_ based on hardware.

(cf https://github.com/flightaware/dump978/blob/master/soapy_source.cc#L239 as a similar thing in dump978's soapysdr code)

File: sdr.c [hunk 4]
```diff
@@ -128,6 +131,9 @@ static sdr_handler sdr_handlers[] = {
 #ifdef ENABLE_LIMESDR
     { "limesdr", SDR_LIMESDR, limesdrInitConfig, limesdrShowHelp, limesdrHandleOption, limesdrOpen, limesdrRun, noStop, limesdrClose, noGetGain, noGetMaxGain, noGetGainDb, noSetGain },
 #endif
+#ifdef ENABLE_SOAPY
+    { "soapy", SDR_SOAPY, soapyInitConfig, soapyShowHelp, soapyHandleOption, soapyOpen, soapyRun, noStop, soapyClose, noGetGain, noGetMaxGain, noGetGainDb, noSetGain },
+#endif
```
Comment (mutability) [inline sdr.c:136]: It would be nice to have the gain API implemented here

**Review by mutability — COMMENTED** [review]
Haven't fully reviewed this yet, but it generally looks good so far, thanks!

File: sdr_soapy.c (hunk 1 above)
Comment (bkoehm) [reply sdr_soapy.c:135]: I was anticipating needing it for RTLSDR or RSPduo specific behavior but you're right, so far haven't needed it, so I went ahead and removed it,  (Although I am still wondering about setting gain for RTLSDR.  I'll refer back to that in my upcoming reply to your gain API comment.)

File: sdr_soapy.c (hunk 2 above)
Comment (bkoehm) [reply sdr_soapy.c]: Yes, it's a floating point.  I've removed the divide-by-10 but more on this in my reply regarding the gain API implementation.

File: sdr_soapy.c (hunk 3 above)
Comment (bkoehm) [reply sdr_soapy.c]: I've added the --bandwidth option.

File: sdr.c (hunk 4 above)
Comment (bkoehm) [reply sdr.c]: I've made an attempt at implementing this but I'm not sure I understand how to handle stepping for devices like RTLSDR that has 3dB stepping?  My device seems happy with 1dB stepping so my implementation works for me but I realize it may fail for 3dB steps.  My SDRPlay device has two gain elements: IFGR and RFGR.  When I call `SoapySDRDevice_setGain()` with a gain of 7dB, IFGR is set to 27dB because IFGR minimum is 20dB.  This can be seen by calling `SoapySDRDevice_getGainElement()` for `IFGR`.  I don't know what kind of gain elements (reported by `SoapySDRDevice_listGains()`) devices like RTLSDR has and how `SoapySDRDevice_setGain()` affects them.

I've also modified the output to report the values of gain elements so that someone with a RTLSDR or other device can start the program and get information on what SoapySDR has for their device.  For example, here's what I see for `--gain 7`:
```
soapy: hardware key is RSPdx
soapy: available antennas: Antenna A, Antenna B, Antenna C
soapy: gain set to 7.0dB, IFGR=27.0dB, RFGR=0.0dB
```

That last line is showing gain of 7 because I started with `--gain 7` and IFRGR is showing 27dB because of IFGRmin+gain = 20dB+7dB.  These are the numbers reported by `SoapySDRDevice_getGain()` and `SoapySDRDevice_getGainElement()` after calling `SoapySDRDevice_setGain()` with a value of 7.

**Comment by bkoehm** [comment]
I've also added --rfgain option for SDRPlay devices once I realized these devices have two gain elements.  `--gain` affects the IFGR element and `--rfgain` affects RFGR.

File: sdr.c (hunk 4 above)
Comment (mutability) [reply sdr.c]: I had a detailed comment here but then github ate it :/

tl;dr: call `SoapySDRDevice_getGainRange()` and use the step size in the returned range structure. i.e. you will have a max gain step index of `(range.maximum - range.minimum) / range.step` and for gain index X the gain in dB is `range.minimum +X * range.step`. Beware of `step == 0`, you probably want to force the step size to e.g. 1dB in that case. Let the SoapySDR driver worry about allocating the requested gain between gain elements.

You might want options to (a) let the user control initial gain per gain element (a generalization of your `--rfgain` option) and (b) to make the gain control API control a specific gain element only, rather than all elements.

Comment (bkoehm) [reply sdr.c]: I've changed `soapyGetMaxGain()` to
`(range.maximum - range.minimum) / (!range.step ? 1 : range.step)`
and `soapyGetGainDb()` to
`(range.minimum + step) * (!range.step ? 1 : range.step)`

Regarding your last paragraph about generalizing, you're proposing enhancing the  `--gain` parameter to accept something like `--gain IFGR:x,RFGR:y`?

**Comment by mutability** [comment]
Hey! Thanks for continuing to poke at this. I haven't forgotten about it, just been busy with other stuff -- I hope I can get this merged soon.

**Comment by mutability** [comment]
I finally had a chance to properly review / tweak this and it's been merged to the dev branch.

I made quite a few changes along the way -- can you check that it still looks OK on a sdrplay?

**Comment by bkoehm** [comment]
@mutability Rebuilt with the dev branch and it is working as expected on my RSPdx.  Here is what I use:
```dump1090 --device-type soapy --interactive --interactive-show-distance --gain 7 --lat <lat> --lon <lon>```

```
soapy: selected device:  driver=sdrplay, label=SDRplay Dev0 RSPdx redacted, serial=redacted
[INFO] devIdx: 0
[INFO] SerNo: redacted
[INFO] hwVer: 4
[INFO] rspDuoMode: 0
[INFO] tuner: 1
[INFO] rspDuoSampleFreq: 0.000000
soapy: hardware info:  sdrplay_api_api_version=3.070000, sdrplay_api_hw_version=4
soapy: driver key:      SDRplay
soapy: hardware key:    RSPdx
soapy: total gain:      7.0dB; IFGR=27.0dB; RFGR=0.0dB
soapy: frequency:       1090.0 MHz
soapy: sample rate:     2.4 MHz
soapy: bandwidth:       5.0 MHz
soapy: AGC mode:        manual
soapy: antenna:         Antenna A
soapy: freq correction: 0.0 ppm
[INFO] Using format CS16.
```

Thank you!
